ADBDEV-4603: Allow multiple params in the child nodes of <dxl:TestExpr>

mos65o2 opened the pull request:
Allow multiple params in the child nodes of \<dxl:TestExpr> (#1140)

Problem definition:
When processing a query with a subquery (example in the test), the "Attribute
number ... not found in project list" error could be output. This happened in
the DXL to Plan Statement stage because the attribute was not found in the
list of column references (CMappingColIdVar) when processing the left node of
\<dxl:TestExpr>. \<dxl:TestExpr> is part of the \<dxl:SubPlan> node and the
TestExpr acts as a filter when a tuple is returned from SubPlan at execution
(see ExecScanSubPlan()).
\<dxl:TestExpr> is processed by TranslateDXLSubplanTestExprToScalar(). The
left node was processed depending on the flag "has_outer_refs". If it was true,
the TranslateDXLTestExprScalarIdentToExpr() was called to read one \<dxl:Ident>
and then convert it to a param. If it was false, the function of handling the
DXL tree was called without taking the param into account. In our case, the
"has_outer_refs" flag was false.
"has_outer_refs" was defined at the DXL generation stage in
PdxlnQuantifiedSubplan(). The params of TestExpr are the common columns of
the scalar and the inner part of PhysicalCorrelated***NLJoin node (see the
physical plan).
As it turns out, the left node can contain more than one param, and at this
stage we need to save the list of attributes that act as TestExpr params,
because in further processing we will not be able to determine them. So
instead of a flag "has_outer_refs", we need to form a list of TestExpr params.
However, if I fixed the determination of the TestExpr params, another error
occurred - "Unsupported subplan test expression", since the left node
contained a tree, but TranslateDXLSubplanTestExprToScalar() only supported
id and cast(id). Also, this function read the TestExpr param directly from
the \<dxl:Ident> node (without using CMappingColIdVar).
Therefore, the tree processing function (TranslateDXLToScalar) is used to
process the left node of TestExpr. In order for this function to correctly
process the TestExpr params, the corresponding list of column references
(CMappingColIdVar) is initialized.

Problem summary:
The \<dxl:TestExpr> handler could output errors:
"Attribute number ... not found in project list" and "Unsupported subplan
test expression".
The first error occurred if a reference to an attribute processed as a TestExpr
param (don't confuse with subplan params \<dxl:ParamList>) was not found. This
occurred due to incorrect detection of the presence of TestExpr params
(m_outer_param flag) at the DXL generation stage in the
PdxlnQuantifiedSubplan function and the lack of adding TestExpr params to the
list of references to attributes (CMappingColIdVar) at the \<dxl:TestExpr>
transforming stage.
The second error occurred if the left node \<dxl:TestExpr> contained a tree with
the TestExpr param, since only id and cast(id) were supported (the param was
read directly from the node).
See the minidump for more information, but keep in mind that the TestExpr param
presence flag (m_outer_param) is not serialized.
Support for one param in the left node of TestExpr was added in ac6ce1a.

Possible solutions:
The "Unsupported subplan test expression" error is fixed by using the
TranslateDXLToScalar function, which allows processing the tree in the child
nodes of TestExpr (mostly the left node). But in this case, it is impossible
to get the TestExpr param (currently it is read directly from node). It is
impossible to determine whether this is a param or a Var (there are no
markers indicating this), if there is more than one attribute in the
tree - this will cause the error "Attribute number ... not found in project
list". Therefore, the TestExpr params must be obtained in a separate structure
 and added in advance to the list of references to attributes
(CMappingColIdVar). Determinating the TestExpr params and adding them to the
CMappingColIdVar before handling DXL fixes the "Attribute number ... not
found in project list" error, but since the left node may contain a tree,
this will cause the "Unsupported subplan test expression" error. Therefore,
these errors must be fixed simultaneously.

Solution:
1) Determine the necessary columns as TestExpr params at the DXL generating
stage;
2) Add references to attributes processed as TestExpr params at the DXL to
Plan Statement stage (prefilling CMappingColIdVar);
2) Add support for processing a tree with params in the left node of TestExpr.

Changes:
At the DXL generation stage (in the PdxlnQuantifiedSubplan function),
the determination of the TestExpr params for PhysicalCorrelated***NLJoin
nodes has been unified. Instead of the "has_outer_refs" flag, the presence
of TestExpr params is now determined by the array of columns -
m_test_expr_params (passed via the CDXLScalarSubPlan class, not serialized).
At the DXL to Plan Statement stage in the subplan handler, a copy of the
mapping (list of references to columns) of the subplan is created to
handle \<dxl:TestExpr>, including the TestExpr params (obtained from
m_test_expr_params). We use the copy because we do not want the TestExpr
params to be mixed with the subplan params.
In the \<dxl:TestExpr> handler, the TranslateDXLToScalar function is
used instead of TranslateDXLTestExprScalarIdentToExpr for the left node,
which allows processing trees.
For the right node \<dxl:TestExpr>, TranslateDXLToScalar is also used for
unification purposes. The TranslateDXLTestExprScalarIdentToExpr function
has been removed as unnecessary.

Ticket: ADBDEV-4603

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87557

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87567

mos65o2 committed:
Fix attribute not found in project list with ANY

Problem:
Subqueries using test expr may result in an "attribute not found in the
projection list" error.

Solution:
Define the necessary parameters and support nested processing of the test
expr tree.

Changes:
Adds parameter definition for test expression in exprToDXL step for
CorrelatedInLeftSemiNLJoin. Reworked test expr handler in DXLtoPlStmt step.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/89838

mos65o2 committed:
Add testcase

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/89858

mos65o2 commented:
I want to rework getting columns from the projection node and assigning mappings as it seems redundant now.

mos65o2 committed:
Simplify getting params from Result node

Added a separate function to get projected columns from the Result node.
PdxlRestictRestul returned to its original state.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90258

mos65o2 committed:
Add testcase

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90283

silent-observer commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -702,21 +702,68 @@ CTranslatorDXLToScalar::TranslateDXLScalarSubplanToScalar(
+		dxlop->GetDxlTestExpr(), slink, &test_expr_var_mapping);
 
 	const CDXLColRefArray *outer_refs = dxlop->GetDxlOuterColRefsArray();
-
```
Unrelated line removed

mos65o2 replied:
Done.

silent-observer replied:
It was still removed

mos65o2 replied:
Please look at the changes.

silent-observer commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -2020,6 +1962,7 @@ CTranslatorDXLToScalar::TranslateDXLScalarIdentToScalar(
 				dxlop->GetDXLColRef()->Id()))
 	{
 		// not an outer ref -> Translate var node
+
```
Unrelated line added

mos65o2 replied:
Done.

silent-observer commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -3089,7 +3089,7 @@ CTranslatorExprToDXL::PdxlnRestrictResult(CDXLNode *dxlnode,
 //	@doc:
 //		Helper to build a Result expression with project list
-//		restricted to required columns
+//		restricted to required columns.
```
Unrelated comment changed

mos65o2 replied:
Done.

silent-observer commented on src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLScalarSubPlan.h:
```diff
@@ -61,20 +62,25 @@ class CDXLScalarSubPlan : public CDXLScalar
 	EdxlSubPlanType m_dxl_subplan_type;
 
 	// test expression -- not null if quantified/existential subplan
+
```
Unrelated line added

mos65o2 replied:
Done.

silent-observer commented on src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLScalarSubPlan.h:
```diff
@@ -61,20 +62,25 @@ class CDXLScalarSubPlan : public CDXLScalar
 
 public:
 	// ctor/dtor
+
```
Unrelated line added

mos65o2 replied:
Done.

silent-observer commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,32 @@ drop table tl1;
 drop table tl3;
 drop table tl4;
+
+-- Check deep tree support in Test expression node when outer params are present with =ANY or IN queries
```
Comment mentions =ANY queries, but only IN queries are actually tested. Maybe add tests for =ANY?

mos65o2 replied:
I indicated in the comment that these are equivalent operators. They generate the same physical plan.

silent-observer commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,32 @@ drop table tl1;
+ANALYZE test_tbl_t;
+
+EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM test_tbl_t WHERE
+  (int_t IN (SELECT int_d FROM test_tbl_d)) IN (SELECT int_p = 1 from test_tbl_p) ORDER BY int_t;
```
Why do we need this second part of the test? It produces a slightly different plan without the patch, but the plan seems valid and no GPORCA errors are present.
```
                                    QUERY PLAN                                    
----------------------------------------------------------------------------------
 Sort
   Output: test_tbl_t.text_t, test_tbl_t.int_t
   Sort Key: test_tbl_t.int_t
   ->  Result
         Output: test_tbl_t.text_t, test_tbl_t.int_t
         Filter: (SubPlan 2)
         ->  Gather Motion 3:1  (slice1; segments: 3)
               Output: test_tbl_t.text_t, test_tbl_t.int_t
               ->  Seq Scan on public.test_tbl_t
                     Output: test_tbl_t.text_t, test_tbl_t.int_t
         SubPlan 2
           ->  Result
                 Output: true
                 Filter: (((hashed SubPlan 1)) = ((test_tbl_p.int_p = 1)))
                 ->  Result
                       Output: (test_tbl_p.int_p = 1), (hashed SubPlan 1)
                       ->  Materialize
                             Output: test_tbl_p.int_p
                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                   Output: test_tbl_p.int_p
                                   ->  Seq Scan on public.test_tbl_p
                                         Output: test_tbl_p.int_p
                       SubPlan 1
                         ->  Result
                               Output: test_tbl_d.int_d
                               ->  Materialize
                                     Output: test_tbl_d.int_d
                                     ->  Gather Motion 3:1  (slice3; segments: 3)
                                           Output: test_tbl_d.int_d
                                           ->  Seq Scan on public.test_tbl_d
                                                 Output: test_tbl_d.int_d
 Optimizer: GPORCA
(32 rows)
```

mos65o2 replied:
The second case checks \<dxl:TestExpr>, which contains not only a param, but also a var. It is necessary to make sure that the contexts are [filled correctly](https://github.com/arenadata/gpdb/blob/e428db925c8c478c53d71ab88a4a43b3ac3feb7b/src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp#L713-L757). I added the comment in the test.

mos65o2 replied:
Since then the test has changed, now it does not contain Explain, but contains DXL.

silent-observer commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,32 @@ drop table tl1;
+CREATE TABLE test_tbl_p (text_p text, int_p int) DISTRIBUTED BY (int_p);
+INSERT INTO test_tbl_p VALUES ('0', 0), ('0', 1);
+
+EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM test_tbl_t WHERE
```
It is not clear that falling back to Postgres optimizer is a fail condition, maybe add a comment that GPORCA should be able to handle this successfully? Or use `set client_min_messages = log` the way it's used in the ticket, so that the error is clearly visible in the diff

mos65o2 replied:
Added GUCs to see the explicit work of the ORCA optimizer.

andr-sokolov commented:
Typo: "physical plane"

mos65o2 committed:
Code refactoring

Removed lines;
Added "optimizer_trace_fallback" in test;
Changed test output.

mos65o2 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4603

mos65o2 committed:
Unify the projection of the inner node of CorrelatedNLJoin

Previously, the projection of the internal node was redefined only for
EopPhysicalCorrelatedLeftOuterNLJoin, but not for
EopPhysicalCorrelatedInLeftSemiNLJoin and
EopPhysicalCorrelatedNotInLeftAntiSemiNLJoin. This was due (most likely)
to the fact that only for EopPhysicalCorrelatedLeftOuterNLJoin the
<dxl:TestExpr> handler could correctly process the left node of
<dxl:TestExpr> with one parameter.
Since the TestExpr handler can now support trees in the left node with
multiple parameters, the projection can be unified for the rest of the CorrelatedNLJoins.
The logic for determining the projection of the columns of the internal node remains the
same. Assert with the dxl_subplan_type definition remains the same.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92385

mos65o2 committed:
Change comments and correct typo

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92429

mos65o2 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4603

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92502

mos65o2 committed:
Add GPOS_DEBUG for variables in ASSERT

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92514

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,37 @@ drop table tl1;
+-- Check deep tree support with params in <dxl:TestExpr> node. TestExpr present with IN queries (equivalent =ANY).
+SET client_min_messages='log';
+SET optimizer_trace_fallback=on;
+CREATE TABLE test_tbl_t (text_t text, int_t int) DISTRIBUTED BY (int_t);
+INSERT INTO test_tbl_t VALUES ('0', 0);
```
what about "create table as select" to reduce queries number?

mos65o2 replied:
Done.

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,37 @@ drop table tl1;
 drop table tl4;
+
+-- Check deep tree support with params in <dxl:TestExpr> node. TestExpr present with IN queries (equivalent =ANY).
+SET client_min_messages='log';
```
The queries in the code above use lower case. The new code style should be the same as style of the code around.

mos65o2 replied:
Done.

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,37 @@ drop table tl1;
+ANALYZE test_tbl_t;
+
+-- After adding data, the ORCA physical plan changes.
+EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM test_tbl_t WHERE
+  (int_t IN (SELECT int_d FROM test_tbl_d)) IN (SELECT int_p = 1 FROM test_tbl_p) ORDER BY int_t;
```
Why do we need two explains? Can we use one explain only? Yes, the plan changes, but why does it matter?

mos65o2 replied:
Explain is needed to understand which DXL was generated. The Explain output does not show that the subquery contains TestExpr (only visible in DXL). But the output shows that the DXLs are different. That is, the 2 outputs prove that the TestExpr handler is tested with different <dxl:TestExpr> contents. Added comments in the test.

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -876,73 +870,20 @@ CTranslatorDXLToScalar::TranslateDXLSubplanTestExprToScalar(
-		// test expression is used for non-scalar subplan,
-		// first arg of test expression must be an EXEC param1 referring to subplan output
-		args = gpdb::LAppend(args, outer_expr);
+	outer_expr = TranslateDXLToScalar(outer_child_node, colid_var);
```
```suggestion
	Expr *outer_expr = TranslateDXLToScalar(outer_child_node, colid_var);
```
Now we assign NULL to the outer_expr and inner_expr variables, and then rewrite them. Do we really need to assign NULL?

mos65o2 replied:
NULL initialization is optional. Done.

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -702,21 +702,67 @@ CTranslatorDXLToScalar::TranslateDXLScalarSubplanToScalar(
+						colid, dxl_to_plstmt_ctxt->GetNextParamId(), mdid,
+						type_modifier);
+#ifdef GPOS_DEBUG
+				BOOL is_inserted =
```
How is the is_inserted variable used?

mos65o2 replied:
I forgot to add an assert after, as I did with the subplan params.
We check if the attribute was added to the parameters context. If false is returned, it means that
a param with that ColId is already in the context. We pre-assign the params an id, which will be returned during processing.
It is invalid if we pre-assign an id to the param, and then try to add it to the context, since it was already assigned an id (the ids can be different). But this is unlikely, since subplan params should not be used inside TestExpr (for this we create a new mapping and context).

mos65o2 committed:
Edit test, add assert and remove initialization

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,44 @@ drop table tl1;
+select * from test_tbl_t where
+  (int_t in (select int_d from test_tbl_d)) in (select int_p = 1 from test_tbl_p) order by int_t;
+
+insert into test_tbl_t values (1);
+insert into test_tbl_d values (1);
+insert into test_tbl_p values (0);
+analyze test_tbl_t;
```
We analyze test_tbl_t only. Do we really need to insert data into test_tbl_d and test_tbl_p?

mos65o2 replied:
As far as I remember the plan doesn't change unless you add data to the other tables. Also, I wanted to make sure it outputs multiple rows.

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,44 @@ drop table tl1;
 drop table tl4;
+
+-- Check support <dxl:TestExpr> node. TestExpr present with IN queries (equivalent =ANY).
+create table test_tbl_t as select 0 as int_t;
+create table test_tbl_d as select 0 as int_d;
+create table test_tbl_p as select int_p from (values (0), (1)) as s(int_p);
```
What do the tables names suffixes (t, d, p) mean? If nothing, can we use names like t1, t2, t3? The same question about int_{t,d,p}

mos65o2 replied:
I intuitively took the table names from the ticket. If it is more convenient, let's rename them.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92815

mos65o2 committed:
Output DXL in test and remove line

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92921

RekGRpth commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,68 @@ drop table tl1;
+set optimizer_minidump=always;
+-- end_ignore
+select * from t1 where
+  (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;
+-- start_ignore
+reset optimizer_minidump;
+-- end_ignore
+\! cut -b 18615-33305 $MASTER_DATA_DIRECTORY/minidumps/* | sed 's/></>\n</g' | sed '/Cost\|Properties\|ValuesList/d' | sed '323,344d;79,299d;68,76d;48,53d'
+
+insert into t1 values (1);
+analyze t1;
+-- After adding data, the ORCA physical plan changes and DXL changes too.
+-- The outer <dxl:TestExpr> is empty. The second <dxl:TestExpr> (nested) does not contain a
+-- deep tree. The left node contains an attribute (Ident ColId="0") that is not present
+-- in the inner <dxl:SubPlan> node. Such attributes in <dxl:TestExpr> are treated as Vars.
+-- The previous case contained only params (attributes contained in the
+-- inner <dxl:SubPlan> node). This test aims to verify that <dxl:TestExpr> with params and
+-- Vars simultaneously is handled correctly during the DXL to Plan Statement stage.
+
+-- Removed from DXL: <dxl:Materialize> inside the nested subplane; <dxl:GatherMotion> inside
+-- nested subplan; scan table t from outer <dxl:GatherMotion>.
+\! rm -rf $MASTER_DATA_DIRECTORY/minidumps
+-- start_ignore
+set optimizer_minidump=always;
+-- end_ignore
+select * from t1 where
+  (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;
+-- start_ignore
+reset optimizer_minidump;
+-- end_ignore
+\! cut -b 18031-26444 $MASTER_DATA_DIRECTORY/minidumps/* | sed 's/></>\n</g' | sed '/Cost\|Properties\|ValuesList/d' | sed '156,201d;121,150d;71,107d'
+
+\! rm -rf $MASTER_DATA_DIRECTORY/minidumps
+-- start_ignore
+reset client_min_messages;
+-- end_ignore
+reset optimizer_trace_fallback;
+drop table t1;
+drop table t2;
+drop table t3;
```
If we are changing only the ORCA, then it would be more logical to make a unit test in the mdp-file.

mos65o2 replied:
We are testing the DXL to Plan Statement conversion stage. We are essentially testing the transition between ORCA and GP. Unit tests don't reach this stage. So we have to output DXL in a regression test.

RekGRpth commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,68 @@ drop table tl1;
+-- some projection elements inside <dxl:Aggregate>; scan table t from <dxl:GatherMotion>.
+\! rm -rf $MASTER_DATA_DIRECTORY/minidumps
+-- start_ignore
+set optimizer_minidump=always;
+-- end_ignore
+select * from t1 where
+  (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;
+-- start_ignore
+reset optimizer_minidump;
+-- end_ignore
+\! cut -b 18615-33305 $MASTER_DATA_DIRECTORY/minidumps/* | sed 's/></>\n</g' | sed '/Cost\|Properties\|ValuesList/d' | sed '323,344d;79,299d;68,76d;48,53d'
+
+insert into t1 values (1);
+analyze t1;
+-- After adding data, the ORCA physical plan changes and DXL changes too.
+-- The outer <dxl:TestExpr> is empty. The second <dxl:TestExpr> (nested) does not contain a
+-- deep tree. The left node contains an attribute (Ident ColId="0") that is not present
+-- in the inner <dxl:SubPlan> node. Such attributes in <dxl:TestExpr> are treated as Vars.
+-- The previous case contained only params (attributes contained in the
+-- inner <dxl:SubPlan> node). This test aims to verify that <dxl:TestExpr> with params and
+-- Vars simultaneously is handled correctly during the DXL to Plan Statement stage.
+
+-- Removed from DXL: <dxl:Materialize> inside the nested subplane; <dxl:GatherMotion> inside
+-- nested subplan; scan table t from outer <dxl:GatherMotion>.
+\! rm -rf $MASTER_DATA_DIRECTORY/minidumps
+-- start_ignore
+set optimizer_minidump=always;
+-- end_ignore
+select * from t1 where
+  (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;
+-- start_ignore
+reset optimizer_minidump;
+-- end_ignore
+\! cut -b 18031-26444 $MASTER_DATA_DIRECTORY/minidumps/* | sed 's/></>\n</g' | sed '/Cost\|Properties\|ValuesList/d' | sed '156,201d;121,150d;71,107d'
+
+\! rm -rf $MASTER_DATA_DIRECTORY/minidumps
+-- start_ignore
+reset client_min_messages;
+-- end_ignore
+reset optimizer_trace_fallback;
```
Why couldn't you just do it this way?
```suggestion
explain (costs off)
select * from t1 where (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;

insert into t1 values (1);
analyze t1;

explain (costs off)
select * from t1 where (i1 in (select i2 from t2)) in (select i3 = 1 from t3) order by i1;
```

mos65o2 replied:
Because Explain does not display TestExpr of subplan. You can't see how it changes. As @andr-sokolov noted, it is not so obvious what is being tested.

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorDXLToScalar.cpp:
```diff
@@ -876,73 +870,17 @@ CTranslatorDXLToScalar::TranslateDXLSubplanTestExprToScalar(
 
 	// Translate arguments
 	List *args = NULL;
-
```
Why was the line deleted?

mos65o2 replied:
Done.

mos65o2 committed:
Rewrite test and add lines

mos65o2 committed:
Remove tab from empty line

mos65o2 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4603

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,77 @@ drop table tl1;
 drop table tl4;
+
+-- Check support <dxl:TestExpr> node. TestExpr present with IN queries (equivalent =ANY).
+drop table if exists t1, t2, t3;
```
`drop table if exists` queries should be between `-- start_ignore` and `-- end_ignore`, because existence of the tables should not affect the test results.

andr-sokolov commented on src/test/regress/expected/subselect_optimizer.out:
```diff
@@ -1235,3 +1235,640 @@ drop table tl1;
+\! sed -i 's/Mdid=".*"//' $MASTER_DATA_DIRECTORY/minidumps/dump.xml
+\! sed -i 's/dxl://' $MASTER_DATA_DIRECTORY/minidumps/dump.xml
+-- end_ignore
+select xpath ('.//Plan', xmlparse(document pg_read_file('minidumps/dump.xml')));
```
Can we use PL/pgSQL to indent output of xpath? It is difficult to read and understand the current version of the xml.

RekGRpth commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,77 @@ drop table tl1;
+\! sed -i 's/dxl://' $MASTER_DATA_DIRECTORY/minidumps/dump.xml
+-- end_ignore
+
+select xpath ('.//Plan', xmlparse(document pg_read_file('minidumps/dump.xml')));
```
Is it possible to come up with some unit test for this (not MDP test)?

mos65o2 committed:
Decorate xml out

mos65o2 committed:
Modify indent dxl output and fix analyze

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -614,3 +614,89 @@ drop table tl1;
+-- Statement stage. With the Postgres optimizer, this check does not matter.
+-- This testcase aims to verify that <dxl:TestExpr> with a deep tree and params in the
+-- left node is handled correctly during the DXL to Plan Statement stage.
+-- Note the <Ident> with ColId 26 and 28. This appears from an internal subplan node.
```
May be "Рay attention to the Ident nodes with ColId 26 and 28. They appear from an internal subplan node." ?

andr-sokolov commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -3193,39 +3235,37 @@ CTranslatorExprToDXL::PdxlnQuantifiedSubplan(
+		GPOS_NEW(m_mp) CColRefSet(m_mp, *pexprInner->DeriveOutputColumns());
+	CColRefSet *pcrsScalarUsed = pexprScalar->DeriveUsedColumns();
+	pcrsInner->Intersection(pcrsScalarUsed);
+	GPOS_ASSERT(pcrsInner->Size() > 0);
```
It seems that pcrsInner contains the same data as pcrsUsed in the deleted code. Why is `if` replaced with `GPOS_ASSERT`? Why is size=0 now considered an error, but this case was considered a normal situation before?

andr-sokolov commented:
Typos: Statment, defenition

mos65o2 committed:
Restore the original algorithm for filling pcrsInner

mos65o2 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4603
